tartufo/scanner.py

```diff
@@ -247,6 +265,35 @@ def signature_is_excluded(self, blob: str, file_path: str) -> bool:
             in self.global_options.exclude_signatures
         )
 
+    @staticmethod
+    def rule_matches(rule: Rule, string: str, path: str) -> bool:
```

Comment:
I think this method is probably worth throwing an lru cache around, similar to `should_scan`. I could see it being called many times for the same patterns, so that could save a lot of extra computation.

Reply:
Using lru is causing the tests to fail.

Reply:
@tarkatronic not sure if you have any suggestions

Reply:
This came down to 2 oddities.
1. Apparently the order of decorators mattered. With `lru_cache` before `staticmethod`, I got errors stating:
   > TypeError: the first argument must be callable

   I didn't even want to deal with that one. Then with `staticmethod` before `lru_cache` I got errors stating:
   > TypeError: unhashable type: 'Rule'

   So much easier to deal with!
2. With the second error, I just needed to make `Rule` hashable. This is done by way of adding a "magic method" called `__hash__` as a way to uniquely identify individual rules.